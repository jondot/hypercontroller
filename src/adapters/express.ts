import wrap from 'async-express-mw'

import { ServerAdapter } from '../server'

class ExpressAdapter implements ServerAdapter {
  express: any

  constructor(express: any) {
    this.express = express
  }
  // eslint-disable-next-line
  createApp(opts: any): any {
    return this.express(opts)
  }

  // eslint-disable-next-line
  mountAction(
    _app: any,
    router: any,
    verb: string,
    _parentPath: string,
    path: string,
    middleware: Function[],
    cb: Function
  ) {
    if (middleware && middleware.length > 0) {
      router[verb](path, middleware, wrap(cb))
    } else {
      router[verb](path, wrap(cb))
    }
  }

  // eslint-disable-next-line
  mountController(app: any, router: any, path: string, middleware: Function[]) {
    app.use(path, middleware, router)
  }

  // eslint-disable-next-line
  createRouter() {
    return this.express.Router()
  }
}

export default ExpressAdapter
