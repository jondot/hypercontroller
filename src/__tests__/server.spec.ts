import request from 'supertest'
import express from 'express'
import {
  Server,
  ServerAdapter,
  get,
  Post,
  Controller,
  GetWithRoute,
  Delete
} from '../index'
import ExpressAdapter from '../adapters/express'

function someMiddleware(req: any, _res: any, next: any) {
  req.hello = 'world'
  next()
}

@Controller(someMiddleware)
class TestingController {
  @GetWithRoute('/')
  get(req: any, res: any) {
    res.json({ middleware: req.hello, get: 'bar' })
  }

  @Post()
  foobar(_req: any, _res: any) {}

  @Delete()
  index(_req: any, _res: any) {}

  @get(someMiddleware)
  withMiddleware(_req: any, _res: any) {}

  custom(_req: any, res: any) {
    res.json({ custom: true })
  }

  _routes({ router }: { router: any }) {
    router.get('custom', this.custom)
  }
}

class TestAdapter implements ServerAdapter {
  mountedActions: any[] = []

  mountedControllers: any[] = []

  createApp(_opts: any): any {
    return { app: true }
  }

  mountAction(
    app: any,
    router: any,
    verb: string,
    parentPath: string,
    path: string,
    middleware: Function[],
    cb: Function
  ) {
    this.mountedActions.push({
      app,
      router,
      verb,
      parentPath,
      path,
      middleware,
      cb
    })
  }

  mountController(app: any, router: any, path: string, middleware: Function[]) {
    this.mountedControllers.push({
      app,
      router,
      path,
      middleware
    })
  }

  createRouter() {
    return { router: true, get: () => {} }
  }
}

describe('server', () => {
  it('should mount controllers and actions', () => {
    const testAdapter = new TestAdapter()

    const server = new Server(testAdapter)
    server.mount([new TestingController()])
    expect(testAdapter.mountedActions).toMatchSnapshot('mounted actions')
    expect(testAdapter.mountedControllers).toMatchSnapshot(
      'mounted controllers'
    )
  })
  it('should respond with mounted middleware', async () => {
    const testAdapter = new ExpressAdapter(express)

    const server = new Server(testAdapter)
    server.mount([new TestingController()])
    const res = await request(server.app).get('/testing')
    expect(res.body).toMatchSnapshot()
  })
})
