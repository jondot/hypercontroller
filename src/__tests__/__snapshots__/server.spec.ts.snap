// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server should mount controllers and actions: mounted actions 1`] = `
Array [
  Object {
    "app": Object {
      "app": true,
    },
    "cb": [Function],
    "middleware": Array [],
    "parentPath": "/testing",
    "path": "/",
    "router": Object {
      "router": true,
    },
    "verb": "get",
  },
  Object {
    "app": Object {
      "app": true,
    },
    "cb": [Function],
    "middleware": Array [],
    "parentPath": "/testing",
    "path": "/foobar",
    "router": Object {
      "router": true,
    },
    "verb": "post",
  },
  Object {
    "app": Object {
      "app": true,
    },
    "cb": [Function],
    "middleware": Array [],
    "parentPath": "/testing",
    "path": "/",
    "router": Object {
      "router": true,
    },
    "verb": "delete",
  },
  Object {
    "app": Object {
      "app": true,
    },
    "cb": [Function],
    "middleware": Array [
      [Function],
    ],
    "parentPath": "/testing",
    "path": "/with-middleware",
    "router": Object {
      "router": true,
    },
    "verb": "get",
  },
]
`;

exports[`server should mount controllers and actions: mounted controllers 1`] = `
Array [
  Object {
    "app": Object {
      "app": true,
    },
    "middleware": Array [
      [Function],
    ],
    "path": "/testing",
    "router": Object {
      "router": true,
    },
  },
]
`;

exports[`server should mount controllers and actions: mountpoints 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "middleware": Array [],
        "parent": "/testing",
        "path": "/",
        "verb": "get",
      },
      Object {
        "middleware": Array [],
        "parent": "/testing",
        "path": "/foobar",
        "verb": "post",
      },
      Object {
        "middleware": Array [],
        "parent": "/testing",
        "path": "/",
        "verb": "delete",
      },
      Object {
        "middleware": Array [
          "someMiddleware",
        ],
        "parent": "/testing",
        "path": "/with-middleware",
        "verb": "get",
      },
    ],
    "middleware": Array [
      "someMiddleware",
    ],
    "path": "/testing",
  },
]
`;
